src/helpers/privacy_budget.rs

```diff
@@ -74,23 +63,20 @@ impl PrivacyBudget {
     /// This atomically increases the budget of all of the blinded match keys that are received.
     /// For those that end up exceeding the budget, the budget is then decremented again,
     /// but the corresponding match key is discarded and not passed to the output.
-    pub fn get_eligible_packets_with_budget(
+    pub fn get_eligible_packets_with_budget<'a>(
         &mut self,
-        packets: &[BlindedMatchKey],
+        packets: impl Iterator<Item = &'a BlindedMatchKey>,
         budget: u32,
-    ) -> Vec<BlindedMatchKey> {
-        let mut eligible_packets: Vec<BlindedMatchKey> = Vec::new();
-        for packet in packets.iter() {
-            let _eligible = match self.increase_budget_consumed(packet, budget) {
-                Ok(true) => {
-                    eligible_packets.push(*packet);
-                    Ok(())
-                }
-                Ok(false) => self.decrease_budget_consumed(&*packet, budget),
-                Err(_) => Ok(()),
-            };
-        }
-        eligible_packets
+    ) -> Vec<&'a BlindedMatchKey> {
+        packets
+            .filter(
+                |packet| match self.increase_budget_consumed(packet, budget) {
+                    Ok(true) => true,
+                    Ok(false) => false,
+                    Err(_) => false,
+                },
+            )
+            .collect()
```

Comment:
You shouldn't need to collect, though it means changing the return value to `impl Iterator<Item = &'a BlindedMatchKey` as well.